flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AbstractEpochManager.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    /**
+     * Add a record to the current epoch and return the current open epoch, the epoch will be
+     * associated with the {@link RecordContext} of this record. Must be invoked within task thread.
+     *
+     * @return the current open epoch.
+     */
+    public Epoch onRecord() {
+        Epoch lastEpoch = outputQueue.get(outputQueue.size() - 1);
+        lastEpoch.ongoingRecordCount++;
+        return lastEpoch;
+    }
+
+    /**
+     * Add a non-record to the current epoch, close current epoch and open a new epoch. Must be
+     * invoked within task thread.
+     *
+     * @param action the action associated with this non-record.
+     */
+    public abstract void onNonRecord(Runnable action);
+
+    /**
+     * Complete one record in the specific epoch. Must be invoked within task thread.
+     *
+     * @param epoch the specific epoch
+     */
+    public void completeOneRecord(Epoch epoch) {
+        epoch.ongoingRecordCount--;
+        // If this epoch has been closed before and all records in
+        //  is epoch have finished, the epoch will be removed from the output queue.
+        if (epoch.tryFinish() && outputQueue.size() > 0) {
+            if (epoch == outputQueue.getFirst()) {
+                outputQueue.remove(0);
+            }
+        }
+    }
+
+    enum EpochStatus {
+        /**
+         * The subsequent non-record input has not arrived. So arriving records will be collected
+         * into current epoch.
+         */
+        Open,
+        /**
+         * The records belong to this epoch is settled since the following non-record input has
+         * arrived, the newly arriving records would be collected into the next epoch.
+         */
+        Closed,
+        /** The records of this epoch have finished execution after the epoch is closed. */
+        Finished
+    }
+
+    /**
+     * All inputs are segment into distinct epochs, marked by the arrival of non-record inputs.
+     * Records are assigned to a unique epoch based on their arrival.
+     */
+    public static class Epoch {
+        /** The number of records that are still ongoing in this epoch. */
+        int ongoingRecordCount;
+
+        /** The action associated with non-record of this epoch(e.g. advance watermark). */
+        @Nonnull Runnable action;
+
+        EpochStatus status;
+
+        public Epoch(int recordCount, Runnable action) {
+            this.ongoingRecordCount = recordCount;
+            this.action = action;
+            this.status = EpochStatus.Open;
+        }
+
+        /**
+         * Try to finish this epoch.
+         *
+         * @return whether this epoch has been finished.
+         */
+        boolean tryFinish() {
+            if (ongoingRecordCount == 0 && this.status == EpochStatus.Closed) {
+                this.status = EpochStatus.Finished;
+                action.run();
+                return true;
+            }
+            return false;
```

Comment:
If this method is invoked on an already finished epoch, the result should be true.

Reply:
Once an epoch is finished, it will be removed from the queue, one epoch won't finish twice.